d/source/flexver.d

```diff
@@ -0,0 +1,440 @@
+			if (diff != 0)
+			{
+				return diff;
+			}
+		}
+		else
+		{
+			return firstComponent.compareEmptyComponent;
+		}
+
+		++componentIndex;
+	}
+
+	if (decomposedSecond.length > componentIndex)
+	{
+		return -decomposedSecond[componentIndex].compareEmptyComponent;
+	}
+
+	return 0;
+}
+
+private Component[] decompose(in string rhs) pure @safe //may throw due to autodecoding
+out (r; isValid(r))
+{
+	Component[] result;
+	Component current;
+
+	if (rhs[0].isDigit)
+	{
+		current.type = ComponentType.Numeric;
+	}
+	else
+	{
+		if (rhs[0] == '-')
+		{
+			current.type = ComponentType.PreRelease;
+		}
+		else
+		{
+			current.type = ComponentType.Textual;
+		}
+	}
+
+	foreach (codePoint; rhs.byCodePoint)
+	{
+		if (codePoint == '+')
+		{
+			result ~= current;
+			return result;
+		}
+
+		final switch (current.type)
+		{
+		case ComponentType.Textual:
+			if (!(codePoint.isDigit))
+			{
+				current.s ~= codePoint;
+			}
+			else
+			{
+				result ~= current;
+				current = Component.init;
+				current.type = ComponentType.Numeric;
+				current.s ~= codePoint;
+			}
+			break;
+
+		case ComponentType.Numeric:
+			if (codePoint.isDigit)
+			{
+				current.s ~= codePoint;
+			}
+			else
+			{
+				result ~= current;
+
+				current = Component.init;
+				if (codePoint == '-')
+				{
+					current.type = ComponentType.PreRelease;
+				}
+				else
+				{
+					current.type = ComponentType.Textual;
+				}
+				current.s ~= codePoint;
+			}
+			break;
+
+		case ComponentType.PreRelease:
+			if (!(codePoint.isDigit))
+			{
+				current.s ~= codePoint;
+			}
+			else
+			{
+				if (current.s.length <= 1)
+				{
+					current.type = ComponentType.Textual;
+				}
+				result ~= current;
+				current = Component.init;
+				current.type = ComponentType.Numeric;
+				current.s ~= codePoint;
+			}
+			break;
+
+		case ComponentType.None:
+			assert(false);
+		}
+	}
+
+	result ~= current;
+	return result;
+}
+
+private bool isValid(in Component[] list) pure @safe
+{
+	ComponentType lastType;
+	foreach (componentIndex, Component c; list)
+	{
+		if (!c.isValid)
+		{
+			return false;
+		}
+
+		immutable temp = c.type;
+		if (temp != ComponentType.Textual && temp != ComponentType.Numeric &&
+				temp != ComponentType.PreRelease)
+		{
+			return false;
+		}
+
+		if (componentIndex == 0)
+		{
+			continue;
+		}
+
+		if (temp == lastType)
+		{
+			return false;
+		}
+		lastType = temp;
+	}
+
+	return true;
+}
+
+private bool isValid(Component component) pure @safe
+{
+	if (component.s.isAppendix)
+	{
+		return false;
+	}
+
+	final switch (component.type)
+	{
+	case ComponentType.Textual:
+		break;
+
+	case ComponentType.Numeric:
+		foreach (codePoint; component.s.byCodePoint)
+		{
+			if (!codePoint.isDigit)
+			{
+				return false;
+			}
+		}
+		break;
+
+	case ComponentType.PreRelease:
+		if (!component.s.isPreRelease)
+		{
+			return false;
+		}
+		break;
+
+	case ComponentType.None:
+		return false;
+	}
+
+	return true;
+}
+
+version (unittest)
```

Comment:
Why have unit test utils in a separate block instead of inside of the unittest block itself?

Reply:
not sure. maybe we thought it was going to be used in multiple tests at one point?
we've moved them back inside.